File: Nov2025/src/main/java/com/nykaa/utilities/ActionHelper.java
package com.nykaa.utilities;

import java.io.IOException;
import java.time.Duration;

import org.openqa.selenium.*;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.*;

public class ActionHelper {

    private WebDriver driver;
    private WebDriverWait wait;
    private JavascriptExecutor js;

    //  Constructor
    public ActionHelper(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(25));
        this.js = (JavascriptExecutor) driver;
    }

    // ✅ Wait until element is visible
    public WebElement waitForVisible(By locator) {
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    // ✅ Wait until element is clickable
    public WebElement waitForClickable(By locator) {
        return wait.until(ExpectedConditions.elementToBeClickable(locator));
    }

    // ✅ Scroll into view safely
    public void scrollIntoView(WebElement element) {
        js.executeScript("arguments[0].scrollIntoView({behavior: 'smooth', block: 'center'});", element);
    }

    // ✅ Move to element (hover)
    public void hover(WebElement element) {
        Actions actions = new Actions(driver);
        actions.moveToElement(element).perform();
    }

    // ✅ Click using JS + screenshot + safe sleep
    public void click(By locator, String screenshotName) throws IOException {
        WebElement element = waitForClickable(locator);
        scrollIntoView(element);

        try {
            js.executeScript("arguments[0].click();", element);
            try {
                Thread.sleep(800); // small pause for UI update
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            Screenshots.capture(driver, screenshotName);
            System.out.println("Clicked element: " + locator);
        } catch (Exception e) {
            System.out.println(" Click failed for: " + locator + " -> " + e.getMessage());
            Screenshots.capture(driver, screenshotName + "_Failed");
        }
    }

    // ✅ Type text safely with clear + screenshot
    public void type(By locator, String text, String screenshotName) throws IOException {
        WebElement input = waitForVisible(locator);
        scrollIntoView(input);
        input.clear();
        input.sendKeys(text);

        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        Screenshots.capture(driver, screenshotName);
        System.out.println("⌨Typed text into: " + locator);
    }

    // ✅ Press ENTER key
    public void pressEnter(By locator) {
        WebElement input = waitForVisible(locator);
        input.sendKeys(Keys.ENTER);
    }

    // ✅ Verify element presence (for assertions)
    public boolean isElementPresent(By locator) {
        try {
            wait.until(ExpectedConditions.presenceOfElementLocated(locator));
            return true;
        } catch (TimeoutException e) {
            return false;
        }
    }
}

File: Pycharm_Python-main/Pycharm_Python-main/NykaaPythonNew/utilities/action_helper.py
import os
import time
from datetime import datetime
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains


class ActionHelper:
    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 20)

    def click(self, locator, description=""):
        try:
            element = self.wait.until(EC.element_to_be_clickable(locator))
            element.click()
            print(f"✅ {description}: Clicked {locator}")
        except Exception as e:
            print(f"❌ {description} failed: {e}")
            self.take_screenshot(description)

    def type(self, locator, text, description=""):
        try:
            element = self.wait.until(EC.visibility_of_element_located(locator))
            element.clear()
            element.send_keys(text)
            print(f"✅ {description}: Typed '{text}'")
        except Exception as e:
            print(f"❌ {description} failed: {e}")
            self.take_screenshot(description)

    def press_enter(self, locator):
        try:
            element = self.wait.until(EC.visibility_of_element_located(locator))
            element.send_keys(Keys.ENTER)
            print("✅ Pressed Enter")
        except Exception as e:
            print(f"❌ Press Enter failed: {e}")
            self.take_screenshot("PressEnter")

    def wait_for_visible(self, locator):
        return self.wait.until(EC.visibility_of_element_located(locator))

    def hover(self, element):
        try:
            self.driver.execute_script("arguments[0].scrollIntoView(true);", element)
            time.sleep(0.5)
            ActionChains(self.driver).move_to_element(element).pause(0.5).perform()
            print("✅ Hovered over element")
        except Exception as e:
            print(f"⚠️ Hover failed: {e}")
            self.take_screenshot("HoverFailed")

    def take_screenshot(self, name="screenshot"):
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        folder = "screenshots"
        os.makedirs(folder, exist_ok=True)
        path = os.path.join(folder, f"{name}_{timestamp}.png")
        self.driver.save_screenshot(path)
        print(f"📸 Screenshot saved: {path}")
